optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/DefaultSolverJob.java

```diff
@@ -109,7 +116,20 @@ public Solution_ call() {
 
     @Override
     public void terminateEarly() {
-        solver.terminateEarly();
+        boolean cancelled = future.cancel(false);
+        if (cancelled) {
+            solverStatus = SolverStatus.NOT_SOLVING;
+        } else {
+            // The solver is either actively solving or has already terminated
+            solver.terminateEarly();
+            try {
+                // Don't return until bestSolutionConsumer won't be called any more
+                terminatedLatch.await();
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                logger.warn("The terminateEarly() call is interrupted.", e);
```

Comment:
Yes, I used a warn here. Usually, that's not a good idea.
Motivation:
A) When things go wrong during a normal shutdown, I want it to stick out, so we can figure out what's going on. Doing logger.info or lower means we'll never know.
B) Fail fasting by throwing an exception, might cause more havoc than it's worth? During shutdown, it's fairly normal that ALL threads get interrupted. I 'd presume that the "shutdown thread" isn't getting interrupted (that would be stupid) and that that one will call terminateEarly() (actually, it will call solverManager.shutdown(), so this discussion is void!).

I am fairly convinced of not doing A) (an info or lower log). I am less convinced for B)
 Wdyt? Should we fail fast instead?